Bluetooth: Host: Add LE Power Control Procedure APIs

KyraLengfeld opened the pull request:
This PR aims to define host APIs, events and configs for LE Power Control Feature. The support of feature is provided with controller-based feature selection with BT_CTLR_LE_POWER_CONTROL_SUPPORT. If the feature is supported, then BT_TRANSMIT_POWER_CONTROL feature will be selectable.

With the new APIs, the applications will:

get improved reading of local and remote tx power
be aware of changes in remote or local tx power
be able to request a tx power change on remote device
be able to set automatic Power Control Request when leaving golden range of RSSI
be able to set APR handling of remote apr to use in local tx power setting
read average RSSI

Defined HCI commands in Core Spec v5.3:

7.8.117 LE Enhanced Read Transmit Power Level command: improvement to existing local tx power reading.
7.8.118 LE Read Remote Transmit Power Level command: Remote tx power is read through an event (LE Transmit Power Reporting)
7.8.121 LE Set Transmit Power Reporting Enable command: Enables local or remote tx power reporting to monitor changes in tx power
7.7.65.33 LE Transmit Power Reporting event:

Vendor-specific HCI commands:
Write remote transmit power level
Set LE Power Control Request parameters

KyraLengfeld committed:
Bluetooth: Host: Add LE Power Control Procedure APIs

This is still a prototype, extracted from zephyrproject-rtos#48451

Plan is to add also VS HCI commands, as they are needed for proper
functunality of LE Power Control.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

KyraLengfeld committed:
Bluetooth: Host: Add LE Power Control Procedure APIs VS cmnds

WIP, testing atm.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+ *  @param conn			Connection object.
+ *  @param phy			PHY bit number i.e. [1M, 2M, s8, s2] == [1, 2, 3, 4]
+ *  @param delta        Delta to be used by remote to apply to its transmit power.
+ *                      Delta is what we as local ask of the peer, it is not
+ *                      synonymous with the changes peer will (or won't) apply.
```
This sentence is complicated. Can we just call it a "request" instead?

KyraLengfeld replied:
What do you think of the new sentences?

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+ *  @param delta        Delta to be used by remote to apply to its transmit power.
+ *                      Delta is what we as local ask of the peer, it is not
+ *                      synonymous with the changes peer will (or won't) apply.
+ *                      Furthermore, it can be 0 to inquire tx power setting of remote.
```
How will the result be reported? I think we should note that here.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+int bt_conn_set_remote_tx_power_level(struct bt_conn *conn,
+					 enum bt_conn_le_tx_power_phy phy, int8_t delta);
+
+/** @brief Set remote (peer) transmit power.
```
This has the same description as `bt_conn_set_remote_tx_power_level`. Is that correct?

KyraLengfeld replied:
whoops.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+ *  @param auto_enable       Enable or Disable controller initiated autonomous
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param apr_enable        Enable or Disable APR handling in controller during
```
What's APR?

KyraLengfeld replied:
Added into sentence, let me know if it is clear enough written.

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+ *  @param apr_enable        Enable or Disable APR handling in controller during
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param beta              Beta value used for exponential weighted averaging of
```
Should we link https://en.wikipedia.org/wiki/Exponential_smoothing?

KyraLengfeld replied:
Is it normal to add wiki pages? And how to write it in that case? "see mathematics: wikipedia"? :D

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+ *                           Core_v5.3, Vol 6, Part B, Section 5.1.17.1, in dBm units.
+ *                           Default value is -70 dBm.
+ *  @param upper_limit       The upper limit of RSSI golden range, it is explained in
+ *                           Core_v5.3, Vol 6, Part B, Section 5.1.17.1, in dBm units.
```
The core spec reference is repeated a lot. I suggest we just define "golden range" once in the description.

KyraLengfeld replied:
Or I change it like this. (I _really_ hope the comments pop up same spot on the new commit :D )

alwa-nordic commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -597,6 +597,71 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+ *                           LE Power Control Request procedure.
+ *                           Disabled by default.
+ *  @param beta              Beta value used for exponential weighted averaging of
+ *                           RSSI in 12-bit fixed point fraction.
```
It's not so clear what "12-bit fixed point fraction" means and it's confusing. Can we instead move the text "The beta value used in computation is beta/4096." up here?

KyraLengfeld replied:
Did so but it still seems a bit convoluted to me. Any ideas? I think it is most important to understand the gist of this... and anyways when I looked to understand this all books referencing this weren't available for free online, so anyways a bit difficult to find deeper understanding (and the algorithm was created by someone who left Nordic a while back)

alwa-nordic replied:
We can't document it as "Lost knowledge. Author quit." :P It has to mean something to our users.

I recognize `avg[n] = beta * avg[n - 1] + (1 - beta) * rssi[n]`. It's the mathematical form of the [Exponential Moving Average]. I assume `avg[n]` is then compared against the targets and limits.

It's not clear to me what `rssi[n]`. Is it the RX RSSI value observed at LL PDU number 'n'? Or the connection event number 'n'? Or something else?

'APR' is not a well known acronym. 'Acceptable power reduction'? Let's spell it out like 'Acceptable power reduction (APR)'. And what effect does that parameter have here?

`wait_period_ms` is also not clear to me. 

> Can we instead move the text "The beta value used in computation is beta/4096." up here?

> Did so but it still seems a bit convoluted to me. What is the trigger of the wait timeout?

I meant to suggest that we somehow replace the sentence "Beta value used for exponential weighted averaging of RSSI in 12-bit fixed point fraction." or remove it entirely. I think we should try to phrase the documentation in terms of what it means for a user of this API. I.e. What is the effect of various values for this parameter.

[Exponential Moving Average]: https://en.wikipedia.org/wiki/Exponential_smoothing

KyraLengfeld replied:
Hi, I tried to make it more readable. I left the math, but renamed changed the beta/4096 to "gamma" in the description, so it is easier to understand: https://github.com/nrfconnect/sdk-zephyr/pull/1376/files#diff-4538e90f4e61dd34802aa1c0a5382dc89d4df6032ec6a3dbbf4eef746ebdd488R642
As per Teams communication I opened a PR in sdk-zephyr directly. If you want me to update this one as well, let me know.

alwa-nordic reviewed (commented):
I'm not finished with the review. I'll continue tomorrow. You don't have to reply yet if you prefer me to finish first.

KyraLengfeld committed:
Bluetooth: Host: fix conn.h descriptions

[DNM] as is. This commit will be squashed, only separated for sake of
visibility.

Signed-off-by: Kyra Lengfeld <kyra.lengfeld@nordicsemi.no>

KyraLengfeld commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -479,6 +479,38 @@ struct bt_conn_le_tx_power {
+	 */
+	int8_t tx_power_level_flag;
+
+	/** Change in transmit power level */
```
Let me know if my formatting is good or has to be changed.

KyraLengfeld commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -530,6 +562,106 @@ int bt_conn_get_remote_info(struct bt_conn *conn,
+ *
+ *  @param conn			Connection object.
+ *  @param phy			PHY bit number i.e. [1M, 2M, s8, s2] == [1, 2, 3, 4]
+ *  @param delta        Delta to be used by remote to apply to its transmit power.
```
Transmit power is in dBm, so an absolute value, and we request a delta of that, so the unit of delta is dBm... but might be wrong the sun has gone down hours ago (exaggeration) so my brain might be sleeping :D

alwa-nordic replied:
Oh. Ok. I assumed it was the same kind of 'delta' as in Core 5.4 4.E.7.7.65.33. That has the unit dB.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/48479610/d7811587-c34d-4d03-b46d-690308d0ee7e)
